### PR TITLE
Sync pr-body artifact updates from any patch-agent session

### LINKED ISSUE #215: catch post-feedback updates to patch artifacts and sync them to PR state

### PR DESCRIPTION
**Pull request by subsetpark**
## Summary

- Generalize the post-session pr-body sync so the orchestrator detects artifact changes from any session kind (CI/Review/Human/Merge_conflict/Rebase), not just Pr_body. Pr_body's existing delivery path (unlink → apply → `classify_pr_body_respond` → miss-counter) is preserved unchanged.
- Decompose the sync into three pure decision functions in `lib/patch_decision.ml` (`pr_body_artifact_changed`, `plan_artifact_sync`, `classify_artifact_sync_outcome`) plus a thin effectful handler in `bin/main.ml`. PATCH failures during opportunistic sync are log-only — they do not couple unrelated operation outcomes to PR-body state.
- Fix `test/dune`: multiple `(test ...)` stanzas in one directory were sharing modules, breaking `dune build @check`. Each stanza now declares `(modules <name>)`.

Closes #215.

## Test plan

- [x] `dune build` clean
- [x] `dune build @check` clean (now passes after dune fix)
- [x] `dune runtest` clean
- [x] 14 enumerated cases (AS-1..AS-24) + 5 QCheck2 properties (AS-P1..AS-P5) for the new pure functions
- [x] Interleaving properties (PI-1..PI-AH-5) extended with `Apply_sync_outcome` command and INV-A/B/C invariants — `Sync_delivered_k` ⇒ `pr_body_delivered ∧ miss_count=0`; `Sync_no_op_k`/`Sync_patch_failed_k` ⇒ no state mutation; generator never produces `(Pr_body, Sync_delivered_k|Sync_patch_failed_k)`
- [ ] Manual sanity (deferred): trigger a Human op via `send_human_message`, have the agent write to `Project_store.pr_body_artifact_path`, confirm the PR description updates and the event log shows "pr-body: synced opportunistically from human session"

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/flowglad/codesmith/onton/pr/221"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

---
## Summary by cubic
Opportunistically sync the PR body artifact after any session (CI/Review/Human/Rebase/Merge_conflict), not just `Pr_body`, so PR descriptions stay current. Satisfies #215; `Pr_body` flow is unchanged and opportunistic PATCH failures are log-only.

- **New Features**
  - Detect changes via pre/post snapshots and PATCH only when changed (whitespace-only collapses to None; truncation to empty counts as a change but maps to no‑op via `Empty`). On success set `pr_body_delivered=true` and reset the miss count; otherwise leave state unchanged. Re-derive `session_ok` from the final result before planning.
  - Added pure planner/classifier in `lib/patch_decision` (`pr_body_artifact_changed`, `plan_artifact_sync`, `classify_artifact_sync_outcome`) with a thin handler in `bin/main.ml`; introduced `Orchestrator.reset_pr_body_artifact_miss_count`. Guards: hard-fail if `pr_number` is missing; if the patch has no gameplan entry (e.g., ad‑hoc agent), skip with a log.
  - Tests: 14 enumerated cases + 5 QCheck2 properties, interleaving checks with a new `Apply_sync_outcome` command (INV‑A/B/C). Added a `(Some "x", Some "y")` distinct-content boundary case to AS‑11.

- **Bug Fixes**
  - Fixed `test/dune` module sharing by declaring `(modules <name>)` per test stanza, restoring `dune build @check`.
  - Docs/tests: corrected reachability notes for `Sync_no_op` fallthroughs (`Missing`/`Empty`/`None`) to reflect real paths; no behavior change.

<sup>Written for commit 03338110afc3f72e9c6088ece87d0a12cf2668d8. Summary will update on new commits. <a href="https://cubic.dev/pr/flowglad/onton/pull/221?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Added opportunistic PR body artifact synchronization that detects content changes and syncs them to GitHub even outside of dedicated PR body sessions.

* **Tests**
  * Expanded test coverage for artifact synchronization, including interleaving scenarios and patch decision logic.

### CODE REVIEW

File: bin/main.ml [hunk 1]
```diff
@@ -3464,6 +3476,71 @@ let runner_fiber ~runtime ~env ~config ~project_name ~pr_registry ~transcripts
...
+                                      Patch_decision.plan_artifact_sync ~kind
+                                        ~session_ok ~pre:pr_body_pre_snapshot
+                                        ~post:pr_body_post_snapshot
+                                    in
+                                    let patch_result =
+                                      match plan with
+                                      | Patch_decision.Sync_skip -> None
+                                      | Patch_decision.Sync_attempt_pr_body ->
+                                          let pr =
+                                            Base.Option.value_exn pr_number
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml]: Applied the suggested fix in 944f962. Replaced `Base.Option.value_exn pr_number` with an explicit `match` that raises `failwith` with a descriptive "BUG: Sync_attempt_pr_body reached with no pr_number for <patch_id>" message and a comment documenting the Deliver-arm invariant.

File: test/test_interleaving_properties.ml [hunk 2]
```diff
@@ -829,6 +888,83 @@ let check_human_messages_preserved ~(prev : Patch_agent.t)
...
+    are unchanged from before the command. The runner only logs in those
+    branches; no state mutation. *)
+let check_sync_outcome_invariants ~prev_agents ~curr_orch cmd =
+  match cmd with
+  | Apply_sync_outcome { patch_idx = _; kind; outcome } ->
+      (* INV-C is enforced by the generator: kind=Pr_body always gets
+         Sync_no_op_k. Cross-check here so a regression in the generator
+         surfaces as a test failure rather than silently exercising
+         unreachable runner state. *)
+      (if Operation_kind.equal kind Operation_kind.Pr_body then
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_interleaving_properties.ml:909]: Applied the tighter check in 944f962. The Sync_delivered_k arm now asserts `miss_count = 0` whenever the post-state has `pr_body_delivered = true`, dropping the `not prev.pr_body_delivered` guard so the targeted-already-delivered case is also validated. Comment updated to explain the post-state-only formulation.

File: test/test_interleaving_properties.ml [hunk 3]
```diff
@@ -829,6 +888,81 @@ let check_human_messages_preserved ~(prev : Patch_agent.t)
...
+         | Sync_delivered_k | Sync_patch_failed_k ->
+             failwith
+               "INV-C generator contract violated: Apply_sync_outcome with \
+                kind=Pr_body must yield Sync_no_op_k");
+      List.iter (Orchestrator.all_agents curr_orch)
+        ~f:(fun (a : Patch_agent.t) ->
+          match Map.find prev_agents a.Patch_agent.patch_id with
+          | None ->
+              () (* agent appeared this step — not an Apply_sync_outcome *)
+          | Some (prev : Patch_agent.t) -> (
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_interleaving_properties.ml]: Applied the suggested scoping in fbbd389. INV-A now resolves the targeted pid from `patch_idx` via `resolve_pid patches patch_idx` and looks up only that agent via `Orchestrator.find_agent`, so a bystander left in `delivered=true && miss_count>0` by an earlier command cannot be miscredited as an INV-A violation. `check_sync_outcome_invariants` now takes a `~patches` parameter; the call site at `run_sequence` passes the patches list. INV-B keeps the per-agent iteration since it's intentionally also asserting no cross-agent side effects (the runner only logs in those branches).

File: bin/main.ml [hunk 4]
```diff
@@ -3464,6 +3476,85 @@ let runner_fiber ~runtime ~env ~config ~project_name ~pr_registry ~transcripts
...
+                                                  (Printf.sprintf
+                                                     "BUG: \
+                                                      Sync_attempt_pr_body \
+                                                      reached with no \
+                                                      pr_number for %s"
+                                                     (Patch_id.to_string
+                                                        patch_id))
+                                          in
+                                          let patch =
+                                            Base.List.find_exn
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml]: Applied the suggested fix in fbbd389. Replaced `Base.List.find_exn` with an explicit `match Base.List.find ... with Some p -> p | None -> failwith ...` that carries the patch_id in the message, matching the convention applied to the `pr_number` guard immediately above.

File: bin/main.ml [hunk 5]
```diff
@@ -3464,6 +3476,97 @@ let runner_fiber ~runtime ~env ~config ~project_name ~pr_registry ~transcripts
...
+                                       so this block is a no-op for that
+                                       kind and never double-PATCHes. *)
+                                    let pr_body_post_snapshot =
+                                      read_artifact_file
+                                        (Project_store.pr_body_artifact_path
+                                           ~project_name ~patch_id)
+                                    in
+                                    let plan =
+                                      Patch_decision.plan_artifact_sync ~kind
+                                        ~session_ok ~pre:pr_body_pre_snapshot
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml:3503]: Applied the suggested fix in 3d94810. Re-derived `session_ok` from the (possibly rebound) `result` immediately before `plan_artifact_sync`, with a comment noting the Pr_body_payload arm above can flip result to `Pr_body_miss`. The original `session_ok` binding at line 3439 is still needed for the payload-arm guard, so the new binding shadows it cleanly for the planner call.

File: bin/main.ml [hunk 6]
```diff
@@ -3464,6 +3476,85 @@ let runner_fiber ~runtime ~env ~config ~project_name ~pr_registry ~transcripts
...
+                                              gameplan.Gameplan.patches
+                                              ~f:(fun (p : Patch.t) ->
+                                                Patch_id.equal p.Patch.id
+                                                  patch_id)
+                                          in
+                                          Some
+                                            (apply_pr_body_artifact ~runtime
+                                               ~net ~github ~project_name
+                                               ~patch_id ~pr_number:pr ~patch
+                                               ~gameplan)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml]: Applied the suggested fix in 073d8ff. Sync_attempt_pr_body's patch lookup now handles a missing gameplan entry (ad-hoc agents added via `Orchestrator.add_agent` have a pr_number but no gameplan entry) by logging "pr-body: skipping opportunistic sync — patch has no gameplan entry (likely ad-hoc)" and returning None. `classify_artifact_sync_outcome` already maps Sync_attempt_pr_body+None to Sync_no_op, so the post-classifier behavior is a clean no-op. The `pr_number=None` failwith above is kept since that case really is a runtime invariant — Sync_attempt_pr_body only fires inside the Deliver arm, which requires a PR (and ad-hoc agents do have a pr_number).

File: lib/patch_decision.mli [hunk 7]
```diff
@@ -125,3 +125,71 @@ val classify_pr_body_respond :
...
+    owned by [classify_pr_body_respond] (with its own miss-counter and
+    intervention semantics). [plan_artifact_sync] returns [Sync_skip] for
+    [Pr_body] so the opportunistic block is a no-op for that kind. *)
+
+val pr_body_artifact_changed : pre:string option -> post:string option -> bool
+(** Pure: did the effective body change between the pre- and post-session
+    snapshots? Whitespace-only and empty strings collapse to [None] before
+    comparison, so truncation and whitespace-only writes are not "changes" —
+    matching the existing "Empty → keep initial PR body" branch in
+    [apply_pr_body_artifact]. *)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_decision.mli]: Updated the docstring (and the matching comment in patch_decision.ml) in 073d8ff. The new wording correctly describes the behavior: whitespace-only/empty post collapses to None so a blank post equals no artifact at all, but a truncation from `Some "x"` to `Some ""` differs from pre after normalization (post' = None vs pre' = Some "x"), so it IS a change. The no-op outcome for that case is then delivered defensively by `classify_artifact_sync_outcome` via `apply_pr_body_artifact`'s ```Empty``` result, not by `pr_body_artifact_changed` itself.

File: lib/patch_decision.mli [hunk 8]
```diff
@@ -125,3 +125,71 @@ val classify_pr_body_respond :
...
+    combination so it is safe to apply over arbitrary generators in property
+    tests, even on inputs the runner never produces.
+
+    - [plan = Sync_skip] → [Sync_no_op] regardless of [patch_result].
+    - [plan = Sync_attempt_pr_body], [patch_result = Some `Ok] →
+      [Sync_delivered].
+    - [plan = Sync_attempt_pr_body], [patch_result = Some `Patch_failed] →
+      [Sync_patch_failed].
+    - [plan = Sync_attempt_pr_body], [patch_result = Some (`Missing | `Empty)]
+      or [None] → [Sync_no_op] (defensive — unreachable when the planner gates
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_decision.mli]: Updated the docstring in 39318c9. New wording explains all three fallthrough arms: `Empty (independent emptiness checks in normalize_artifact and apply_pr_body_artifact's String.trim), `Missing (file deleted between post-snapshot and apply call), and None (caller skipped PATCH despite plan — ad-hoc patches with no gameplan entry are one example, added in the previous round of fixes).

File: test/test_patch_decision.ml [hunk 9]
```diff
@@ -751,4 +751,253 @@ let () =
...
+    Stdlib.print_endline "AS-21 passed";
+
+    (* AS-22: plan=Sync_attempt_pr_body, Some `Patch_failed → Sync_patch_failed. *)
+    let outcome =
+      classify_artifact_sync_outcome ~plan:Sync_attempt_pr_body
+        ~patch_result:(Some `Patch_failed)
+    in
+    assert (equal_artifact_sync_outcome outcome Sync_patch_failed);
+    Stdlib.print_endline "AS-22 passed";
+
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_patch_decision.ml:963]: Updated the AS-23 comment in 39318c9 to explain why both arms are reachable: `Empty fires when content goes from non-empty to empty (independent emptiness checks in normalize_artifact and apply_pr_body_artifact's String.trim), `Missing fires if the file is deleted between the runner's post-snapshot and the apply call. The test bodies and assertions are unchanged.

File: test/test_interleaving_properties.ml [hunk 10]
```diff
@@ -829,6 +888,87 @@ let check_human_messages_preserved ~(prev : Patch_agent.t)
...
+             miscredited as an INV-A violation. *)
+          let targeted_pid = resolve_pid patches patch_idx in
+          match Map.find prev_agents targeted_pid with
+          | None -> ()
+          | Some _ -> (
+              match Orchestrator.find_agent curr_orch targeted_pid with
+              | None -> ()
+              | Some (a : Patch_agent.t) ->
+                  if
+                    a.Patch_agent.pr_body_delivered
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_interleaving_properties.ml]: Applied the suggested fix in bfd0f61. INV-A now asserts both halves unconditionally for the targeted agent: `pr_body_delivered = true` (with its own failwith if missing) and `miss_count = 0` (with its own failwith if non-zero). A regression that leaves `pr_body_delivered = false` after `Sync_delivered_k` will now fail loudly instead of passing vacuously via short-circuit.

File: test/test_patch_decision.ml [hunk 11]
```diff
@@ -751,4 +751,257 @@ let () =
...
+          ~f:(fun (pre, post) ->
+            let plan = plan_artifact_sync ~kind ~session_ok:false ~pre ~post in
+            assert (equal_artifact_sync_plan plan Sync_skip)));
+    Stdlib.print_endline "AS-10 passed";
+
+    (* AS-11: kind=Pr_body, session_ok=true → Sync_skip regardless of
+       changed. The Pr_body delivery path is owned by
+       classify_pr_body_respond. *)
+    List.iter
+      [ (None, None); (Some "x", Some "x"); (None, Some "y") ]
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_patch_decision.ml]: Applied the suggested fix in 0333811. Added `(Some "x", Some "y")` — distinct non-empty pre/post — to the AS-11 enumeration so the table self-contains the (Some, Some)-with-content-change case alongside (None, None), (Some, Some)-equal, and None→Some. AS-P4 still covers the contract exhaustively as the safety net.
